src/esw/brushless.py

```diff
@@ -0,0 +1,231 @@
+                    watchdog_timeout=self.WATCHDOG_LIMIT,
+                    query=True,
+                ),
+                timeout=self.TIME_INDICATING_DISCONNECTED,
+            )
+            self._fault_response = state.values[moteus.Register.FAULT]
+        except asyncio.TimeoutError:
+            if self.state != "Disconnected":
+                rospy.logerr(f"CAN ID {self.CAN_ID} disconnected when trying to send command")
+            self.state = "Disconnected"
+
+    def _has_error(self) -> bool:
+        return self._fault_response != 0
+
+    async def _connect(self) -> None:
+        try:
+            await asyncio.wait_for(self.controller.set_stop(), timeout=0.1)
+            await asyncio.wait_for(
+                self.controller.set_position(
+                    position=math.nan,
+                    velocity=0,
+                    velocity_limit=CommandData.VELOCITY_LIMIT,
+                    maximum_torque=CommandData.MAX_TORQUE,
+                    watchdog_timeout=self.WATCHDOG_LIMIT,
+                    query=True,
+                ),
+                timeout=self.TIME_INDICATING_DISCONNECTED,
+            )
+        except asyncio.TimeoutError:
+            if self.state != "Disconnected":
+                rospy.logerr(f"CAN ID {self.CAN_ID} disconnected when trying to connect")
+            self.state = "Disconnected"
+            return
+        self.state = "Armed"
+
+    async def update(self) -> None:
+        if self.state == "Armed":
+            errors = self._has_error()
+
+            if errors:
+                rospy.logerr(f"CAN ID {self.CAN_ID} has error {self._fault_response}")
+                self.state = "Error"
+                return
+
+            await self._send_command()
+
+        elif self.state == "Disconnected":
+            await self._connect()
+
+        elif self.state == "Error":
+            await self.clean_error()
+
+    async def clean_error(self) -> None:
+        try:
+            await asyncio.wait_for(self.controller.set_stop(), timeout=self.TIME_INDICATING_DISCONNECTED)
+            state = await asyncio.wait_for(
+                self.controller.set_position(
+                    position=math.nan,
+                    velocity=0,
+                    velocity_limit=CommandData.VELOCITY_LIMIT,
+                    maximum_torque=CommandData.MAX_TORQUE,
+                    watchdog_timeout=self.WATCHDOG_LIMIT,
+                    query=True,
+                ),
+                timeout=self.TIME_INDICATING_DISCONNECTED,
+            )
+        except asyncio.TimeoutError:
+            if self.state != "Disconnected":
+                rospy.logerr(f"CAN ID {self.CAN_ID} disconnected when trying to clean error")
+            self.state = "Disconnected"
+            return
+        self._fault_response = state.values[moteus.Register.FAULT]
+        if self._has_error():
+            self.state = "Error"
+        else:
+            self.state = "Armed"
```

Comment:
This is very similar to _connect(). Why not just change state to either "Disconnected" or back to "Error" in here? In disconnected state you then call connect and verify the connection anyways. You can check whether the controller receives your "pings" with the return value of `response = await asyncio.wait_for(self.controller.query(), RESPONSE_TIMEOUT)` as seen in my branch. I think set_stop will work too but query() is how moteus actually does it under the hood when looking for controllers. 

Overall not a huge issue but you are doing some extra work here that can other functions are designed to do already in the state machine.

Reply:
Yeah fair point let me try to refactor one sec

Reply:
Ok 
1. you asked why not just change the state to either disconnected or back to error in here (inside clean_error), but it does at the end.
2. basically the purpose of reason why I used set_position instead of query() is because I previously had only set_stop() since that clears errors but then apparently it does not timeout even when the controller is disconnected which is why I had a set_position fx. I'll go into the wc today to test if query() experiences similar behavior but if not, then I'll replace the set_position with query() Either way, I'll rename clear_error to _connect and delete the old _connect.